ElectionGuard-COMP413/src/simple_election_data.ts

```diff
@@ -0,0 +1,1126 @@
+        return votes_cast > 1 ;
+    }
+}
+
+
+export class PlaintextBallotSelection implements OrderedObjectBase {
+    // """
+    // A BallotSelection represents an individual selection on a ballot.
+    // This class accepts a `vote` integer field which has no constraints
+    // in the ElectionGuard Data Specification, but is constrained logically
+    // in the application to resolve to `False` or `True` aka only 0 and 1 is
+    // supported for now.
+    // This class can also be designated as `is_placeholder_selection` which has no
+    // context to the data specification but is useful for running validity checks internally
+    // an `extended_data` field exists to support any arbitrary data to be associated
+    // with the selection.  In practice, this field is the cleartext representation
+    // of a write-in candidate value.  In the current implementation these values are
+    // discarded when encrypting.
+    // """
+
+    // The object id of this specific ballot.
+    // Will also appear in any corresponding plaintext of this ballot.
+    object_id: string;
+
+    sequence_order: number;
+
+
+    //Hash of the election
+
+    vote: number;
+
+    is_placeholder_selection = false;
+    // """Determines if this is a placeholder selection"""
+
+    // # TODO: ISSUE #35: encrypt/decrypt
+    @Type(() => ExtendedData)
+    extended_data?: ExtendedData = undefined;
+    // """
+    // an optional field of arbitrary data, such as the value of a write-in candidate
+    // """
+
+    // 1 implies a vote for. 0 implies no vote.
+    public constructor(object_id: string, sequence_order: number, vote: number, is_placeholder_selection: boolean, extened_data?: ExtendedData){
+        this.object_id = object_id;
+        this.sequence_order = sequence_order;
+        this.vote = vote;
+        this.is_placeholder_selection = is_placeholder_selection;
+        this.extended_data = extened_data;
+    }
+
+    public is_valid(expected_object_id: string): boolean {
+        // """
+        // Given a PlaintextBallotSelection validates that the object matches an expected object
+        // and that the plaintext string can resolve to a valid representation
+        // """
+
+        if (this.object_id !== expected_object_id) {
+            console.log(`invalid object_id: expected(${expected_object_id}) actual(${this.object_id})`);
+            return false;
+        }
+
+        const vote = this.vote;
+        if (vote < 0 || vote > 1) {
+            console.log(`Currently only supporting choices of 0 or 1: ${this.toString()}`);
+            return false;
+        }
+
+        return true;
+    }
+
+    public equals(other: any): boolean {
+        return (
+            other instanceof PlaintextBallotSelection
+            && this.object_id === other.object_id
+            && this.vote === other.vote
+            && this.is_placeholder_selection === other.is_placeholder_selection
+            && this.extended_data === other.extended_data
```

Comment:
So far as I can tell, === is the most intuitive of the various JavaScript equality operators. We want "deep" equality, which is to say, comparing all the fields in any object, rather than "shallow" equality (pointers to the same thing).